### PR TITLE
feat: bulk secondary index builder

### LINKED ISSUE #1005: Add endpoint that builds all indexes in collection
Add an endpoint that will read all documents in collection and build the indexes for it

### PR DESCRIPTION
**Pull request by garrensmith**
## Describe your changes
Adds an endpoint that will read all documents in a collection and add them to the secondary index. The bulk indexer will read in batches. If an error occurs it will try to decrease the number of documents read in a batch and try again.

## How best to test these changes
All tests should pass. C

## Issue ticket number and link
closes #1005

### CODE REVIEW

File: server/services/v1/database/indexer_runner.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	indexer := NewSecondaryIndexer(coll)
+
+	if err = indexer.BuildCollection(ctx, runner.txMgr); err != nil {
+		log.Err(err).Msgf("Failed to index collection \"%s\" for db \"%s\"", coll.Name, db.DbName())
+		return Response{}, ctx, err
+	}
+
+	for _, index := range coll.SecondaryIndexes.All {
```
Comment (himank) [inline server/services/v1/database/indexer_runner.go:58]: What happens if after the BuildCollection req fails or the server restarts, the next retry will trigger the buildCollection from the beginning again?

File: server/services/v1/database/indexer_runner.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	if err = tenant.UpdateCollectionIndexes(ctx, tx, db, coll.Name, coll.SecondaryIndexes.All); ulog.E(err) {
+		return Response{}, ctx, err
+	}
+
+	if err = tx.Commit(ctx); ulog.E(err) {
+		return Response{}, ctx, err
+	}
+
+	runner.queryMetrics.SetWriteType("replace")
```
Comment (himank) [inline server/services/v1/database/indexer_runner.go:75]: Is it safe to use replace here? Or should we have a tag "build_index" here?

File: server/services/v1/database/secondary_indexer.go [hunk 3]
```diff
@@ -134,6 +134,105 @@ func NewSecondaryIndexer(coll *schema.DefaultCollection) *SecondaryIndexer {
...
+		if err != nil {
+			return nil, err
+		}
+		return reader.ScanIterator(from, nil)
+	}
+
+	return reader.ScanTable(table)
+}
+
+var RETRY_ERRORS = []error{
```
Comment (himank) [inline server/services/v1/database/secondary_indexer.go:220]: No need to expose it? Just have `var retryErrors = []error...`

File: server/services/v1/database/secondary_indexer.go [hunk 4]
```diff
@@ -134,6 +134,105 @@ func NewSecondaryIndexer(coll *schema.DefaultCollection) *SecondaryIndexer {
...
+				return err
+			}
+
+			if err = q.Index(ctx, tx, row.Data, fdbKey.IndexParts()); err != nil {
+				return err
+			}
+			count += 1
+		}
+
+		err = tx.Commit(ctx)
```
Comment (himank) [inline server/services/v1/database/secondary_indexer.go:175]: nit, you can combine the below `if` in this line i.e.
```
if err = tx.Commit(ctx); err != nil {
  if !shouldRetryBulkIndex(err) {
    return err
  }
  
  docFetch /= 2
  count = 0
  continue
}
```

File: schema/collection.go [hunk 5]
```diff
@@ -241,6 +241,16 @@ func (d *DefaultCollection) GetActiveIndexedFields() []*QueryableField {
 	return indexed
 }
 
+func (d *DefaultCollection) GetWriteModeIndexes() []*QueryableField {
```
Comment (himank) [inline schema/collection.go:244]: Is this mode for `backfill`?

File: server/services/v1/database/indexer_runner.go (hunk 1 above)
Comment (garrensmith) [reply server/services/v1/database/indexer_runner.go:58]: Yes correct. Once we have background workers we will keep a position and upgrade from there.

File: server/services/v1/database/secondary_indexer.go (hunk 4 above)
Comment (garrensmith) [reply server/services/v1/database/secondary_indexer.go:175]: Nice catch

File: server/services/v1/database/indexer_runner.go (hunk 1 above)
Comment (himank) [reply server/services/v1/database/indexer_runner.go:58]: Sounds good.

File: schema/collection.go (hunk 5 above)
Comment (garrensmith) [reply schema/collection.go:244]: I will use it for the worker indexing

**Review by himank — APPROVED** [review]

**Review by himank — APPROVED** [review]

**Pull request merged**